core/solver/multigrid.cpp

```diff
@@ -159,346 +160,338 @@ void handle_list(
-        r_list.emplace_back(vec::create(exec, dim<2>{current_nrows, nrhs}));
-        g_list.emplace_back(vec::create(exec, dim<2>{next_nrows, nrhs}));
-        e_list.emplace_back(vec::create(exec, dim<2>{next_nrows, nrhs}));
-        one_list.emplace_back(initialize<vec>({gko::one<VT>()}, exec));
-        neg_one_list.emplace_back(initialize<vec>({-gko::one<VT>()}, exec));
-        if ((cycle == multigrid::cycle::kfcg ||
-             cycle == multigrid::cycle::kgcr) &&
-            level % multigrid->get_parameters().kcycle_base == 0) {
-            kcycle_state.allocate_memory<VT>(level, cycle, current_nrows,
-                                             next_nrows);
-        }
-    }
 
+/**
+ * The enum class is to combine the cycle infomation  It's legal to use a binary
+ * or(|) operation to combine several properties.
+ */
+enum class cycle_info {
+    /**
+     * indicate input is zero
+     */
+    x_is_zero = 1,
+
+    /**
+     * current process is the first one of the cycle
+     */
+    first_of_cycle = 2,
+
+    /**
+     * current procees is the end one of the cycle
+     */
+    end_of_cycle = 4
+};
+
+
+GKO_ATTRIBUTES GKO_INLINE cycle_info operator|(cycle_info a, cycle_info b)
+{
+    return static_cast<cycle_info>(static_cast<int>(a) | static_cast<int>(b));
+}
+
+
+GKO_ATTRIBUTES GKO_INLINE bool operator&(cycle_info a, cycle_info b)
+{
+    return static_cast<bool>(static_cast<int>(a) & static_cast<int>(b));
+}
+
+
+/**
+ * MultigridState is to store the necessary cache and run the operation of all
+ * levels.
+ *
+ * @note it should only be used internally
+ */
+struct MultigridState {
+    MultigridState() : nrhs{0} {}
+
+    /**
+     * Generate the cache for later usage.
+     *
+     * @param system_matrix_in  the system matrix
+     * @param multigrid_in  the multigrid information
+     * @param nrhs_in  the number of right hand side
+     */
+    void generate(const LinOp* system_matrix_in, const Multigrid* multigrid_in,
+                  const size_type nrhs_in);
+
+    /**
+     * allocate_memory is a helper function to allocate the memory of one level
+     *
+     * @tparam ValueType  the value type of memory
+     *
+     * @param level  the current level index
+     * @param cycle  the multigrid cycle
+     * @param current_nrows  the number of rows of current fine matrix
+     * @param next_nrows  the number of rows of next coarse matrix
+     */
+    template <typename ValueType>
+    void allocate_memory(int level, multigrid::cycle cycle,
+                         size_type current_nrows, size_type next_nrows);
+
+    /**
+     * run the cycle of the level
+     *
+     * @param cycle  the multigrid cycle
+     * @param level  the current level index
+     * @param matrix  the system matrix of current level
+     * @param b  the right hand side
+     * @param x  the input vectors
+     * @param info  the info of cycle
+     */
     void run_cycle(multigrid::cycle cycle, size_type level,
```

Comment:
Can we rename this to `run_mg_cycle` to avoid confusion with the templated version which this calls ?